presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/execution/ExceptionUtils.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.functionNamespace.execution;
+
+import com.facebook.drift.TException;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.StandardErrorCode;
+import com.facebook.presto.thrift.api.udf.ThriftUdfServiceException;
+import com.facebook.presto.thrift.api.udf.UdfExecutionFailureInfo;
+
+import javax.annotation.Nullable;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+
+public class ExceptionUtils
+{
+    private static final Pattern STACK_TRACE_PATTERN = Pattern.compile("(.*)\\.(.*)\\(([^:]*)(?::(.*))?\\)");
+
+    private ExceptionUtils()
+    {
+    }
+
+    public static PrestoException toPrestoException(ThriftUdfServiceException thriftException)
+    {
+        UdfExecutionException cause = toUdfExecutionException(thriftException.getFailureInfo());
```

Comment:
Is `UdfExecutionException` necessary? It doesn't seem to exist beyond the helper functions.

Reply:
Yes, it is necessary, maybe not obvious, so let me explain.

The goal here is to return and eventually throw a `PrestoException` which includes the full stacktrace that happens on the remote UDF server.

The stacktrace information comes from `UdfExecutionFailureInfo`. We're able to serialize any arbitrary Java exception on the UDF server, and is now trying to deserialize into a Java exception in Presto. However, we won't be able to instantiate the exact same Exception chain in Presto, because exception classes defined in the UDF server binary may not even exists in Presto.

It's not possible to ask Presto to depend on UDF server or to require/enforce UDF servers to only throw Java built-in Exception. Hence, we define this `UdfExecutionException` to mimic the Exception chain structure of the original stacktrace without having to instantiate the specific exception class that is only available in the UDF server binary.

You can see that there is a field `type` in `UdfExecutionException`, which we expect the UDF server to populate with the canonical class name of the original exception class. This allows us the printout of the stacktrace to look very similar.

Reply:
Why can't you directly construct a PrestoException? You can pretty much inline all the logic? Anyways if you want the cause to be set to a UDF specific exception that's fine too. Maybe the class can be defined here in this class since it won't be used anywhere else?

Reply:
Before this PR, we're already throwing a `PrestoException` with GENERIC_INTERNAL_ERROR and no UDF server stacktrace. One purpose of the PR is to show the part of the stacktrace happened on UDF server so that it's complete and allows debugging.

I would prefer the class to be top-level one as it will shows better in the resulting stacktrace printout. Otherwise you'll be seeing `com.facebook.presto.functionNamespace.execution.ExceptionUtils$UdfExecutionException`.

The class mimic exactly [Failure.java](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/execution/Failure.java#L22).